### DISCUSSION TITLE
Localized API calls in PIM 1.6.0 are different to prior version (especially `productAttributValues`)

### DISCUSSION THREAD

**Discussion by marcus-at-localhost** [atrocore/atrocore#597] [Q&A] [answered]
This is related to this issue: https://github.com/atrocore/atrocore/discussions/588 but turns out is something else.

Here is what I do and how the recent changes for localized productAttributValues affect this:

1. Get product info
```
curl -X GET "/api/v1/Product/629fb7ae9625dd7c4" -H  "accept: application/json" -H  "language: en_US" -H  "Authorization-Token: Y"
``` 
=> Response gives me localized entries (`name`, `longDescription` etc)

2. Get all attributes for product
```
curl -X GET "/api/v1/Product/629fb7ae9625dd7c4/productAttributeValues" -H  "accept: application/json" -H  "language: en_US" -H  "Authorization-Token: Y"
``` 
=> **used to return all attributes in the past** now only returns one localized attribute!

3. get explicit localized attribute labels

```
curl -X GET "/api/v1/Attribute?where[0][attribute]=id&where[0][type]=in&where[0][value][0]=60054c5f7aa40c499&where[0][value][1]=60054c5f7adf5c57c&where[0][value][2]=600ee1748caf25e89&where[0][value][3]=600efdf62e4ca5c66&where[0][value][4]=600efdf6374dc3793" -H  "accept: application/json" -H  "language: en_US" -H  "Authorization-Token: Y"
```

 => Returns the localized attribute labels based on the second call.

4. Then merge the result of `productAttributeValues` into the result of `Attributes` and use it further.

With the change in how you deal with localized `productAttributValues` the second step throws all subsequent steps off. To get it back to work, I have to explicitly omit the language header for `productAttributeValues` to get all associated attributes to the product. This kind of breaks the convention to only make calls with a language header.

What would be a better way to get all product related attributes?

I noticed you changed how attributes are loaded in the PIM (per Attribute Group, if I'm not mistaken), but I'd rather have one API call with all results (that I cache) than many small calls (calls are made server side, so I try to have as less calls as possible).

Internally, you don't use the language header anymore, probably to switch languages quicker. Can you please help me to understand the concept of that part of the API/App? 

Thank you so much!

**Comment by rratsun** [comment, marked as answer]
There was a bug here. We are fixed this at PIM 1.6.9.
```curl -X GET "/api/v1/Product/629fb7ae9625dd7c4/productAttributeValues" -H  "accept: application/json" -H  "language: en_US" -H  "Authorization-Token: Y"``` this request should return all language specific attributes and all NOT language specific (main) attributes.

If you want to get only not language specific attributes use ```curl -X GET "/api/v1/Product/629fb7ae9625dd7c4/productAttributeValues" -H  "accept: application/json" -H  "language: main" -H  "Authorization-Token: Y"```

**Reply by marcus-at-localhost** [reply]
Thank you for resolving it so quickly @rratsun  I'm glad it was only a bug and not an API change.
So  the language option 'main' is new and is used to address the PIM's "main" or first language.
That makes sense now, and I believe it is an improvement for using multilanguage attribute values. Excellent work!

With that change in the language setup, is it safe to say that the three language fields are no longer required for a bilingual setup (for example, DE and EN)?


Right now we've got the fields:
- name
- nameDeDe
- nameEnUS

if my main language is "DE" then name and nameDeDe are filled with the same value. That was always a bit confusing for maintainers.

Would you advise to remove nameDeDe then?

It would be great if there was a unified changelog that addressed such changes (I know there's a twitter feed, but it's a little short). I appreciate all of the enhancements and updates; I just get nervous every now and then when I hit the update button:)

**Reply by atrolex** [reply]
As your Main Language is German, you don't need to add / have German as additional language (many users to this mistake, i will think about how / where to declare it).  Make sure that multilingual fields and attributes without locale suffix (Main Language) have a value in German and remove the "de_DE" here: https://demo.atropim.com/#Settings/inputLanguage

At the end of the day you will have only:
- name (<actually German)
- nameEnUs

We plan to introduce a separate subdomain exclusively for release notes, comes very soon. Twitter was already removed :)

Every time you click the "update" button, a backup of your db is done, don't remember where it is placed, but in any case you have there the backups for the last 7 updates. So if something goes wrong, you can recover from it.

**Reply by marcus-at-localhost** [reply]
Thanks @atrolex for clarifying! I guess the misconception of having the same language twice might come from the demo itself? (having main language in English and additional English fields?) And wasn't the main language initially used for listings? At least that's what I remember when I started working with the PIM in early 2020.

Anyway – that's all good to know. Thank you so much!

PS: yes backup sits in the /db-dump folder. I should take the time, setup a second instance and test it. As we all know: Backups are useless if we don't test them :)

**Comment by rratsun** [comment]
@marcus-at-localhost Thank you. We appreciate you to helps us make our product better.

**Reply by marcus-at-localhost** [reply]
Same. You and your team are doing a great job and building a fantastic product! Thank you!
